File: backend/app/schemas/surveys.py
"""问卷 Schema 定义 - 汤吧社区"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime

from app.models.models import SurveyStatus, SurveyQuestionType


class SurveyQuestionCreate(BaseModel):
    question_text: str
    question_type: SurveyQuestionType
    options: Optional[List[str]] = None
    required: bool = True
    sort_order: int = 0


class SurveyCreate(BaseModel):
    title: str = Field(max_length=200)
    description: Optional[str] = None
    status: SurveyStatus = SurveyStatus.DRAFT
    starts_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    questions: List[SurveyQuestionCreate] = []


class SurveyUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = None
    status: Optional[SurveyStatus] = None
    starts_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None


class SurveyQuestionResponse(SurveyQuestionCreate):
    id: int
    survey_id: int
    created_at: datetime

    class Config:
        from_attributes = True


class SurveyResponseBase(BaseModel):
    survey_id: int


class SurveyAnswerCreate(BaseModel):
    question_id: int
    answer_text: Optional[str] = None
    answer_option_ids: Optional[List[int]] = None
    answer_rating: Optional[int] = None


class SurveySubmit(BaseModel):
    answers: List[SurveyAnswerCreate]


class SurveyDetail(SurveyCreate):
    id: int
    author_uid: int
    notification_sent: bool
    created_at: datetime
    updated_at: datetime
    questions: List[SurveyQuestionResponse] = []

    class Config:
        from_attributes = True


class SurveySummary(BaseModel):
    id: int
    title: str
    description: Optional[str] = None
    status: SurveyStatus
    starts_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    notification_sent: bool
    created_at: datetime
    updated_at: datetime
    question_count: int = 0
    response_count: int = 0

    class Config:
        from_attributes = True


class SurveyPageResponse(BaseModel):
    items: List[SurveySummary]
    total: int
    page: int
    page_size: int
    total_pages: int


class SurveyStatistics(BaseModel):
    survey_id: int
    total_responses: int
    question_stats: List[Dict[str, Any]]

File: frontend/src/api/surveys.ts
import http from './http'
import type { PageParams, PageResult } from '@/types'

export interface SurveyQuestion {
  id: number
  survey_id?: number
  question_text: string
  question_type: 'single_choice' | 'multiple_choice' | 'text' | 'rating'
  options?: string[] | null
  required: boolean
  sort_order: number
  created_at?: string
}

export interface Survey {
  id: number
  title: string
  description?: string | null
  status: 'draft' | 'active' | 'closed' | 'expired'
  starts_at?: string | null
  expires_at?: string | null
  notification_sent: boolean
  author_uid: number
  created_at: string
  updated_at: string
  questions?: SurveyQuestion[]
  question_count?: number
  response_count?: number
  has_submitted?: boolean
}

export type SurveyQuestionInput = Omit<SurveyQuestion, 'id' | 'survey_id' | 'created_at'>

export interface SurveyCreate {
  title: string
  description?: string | null
  status?: 'draft' | 'active' | 'closed' | 'expired'
  starts_at?: string | null
  expires_at?: string | null
  questions?: SurveyQuestionInput[]
}

export interface SurveyUpdate {
  title?: string | null
  description?: string | null
  status?: 'draft' | 'active' | 'closed' | 'expired'
  starts_at?: string | null
  expires_at?: string | null
}

export interface SurveyAnswer {
  question_id: number
  answer_text?: string | null
  answer_option_ids?: number[] | null
  answer_rating?: number | null
}

export interface SurveySubmit {
  answers: SurveyAnswer[]
}

export interface SurveyStatistics {
  survey_id: number
  total_responses: number
  question_stats: Array<{
    question_id: number
    question_text: string
    total_answers: number
    option_distribution?: Record<number, number>
    average_rating?: number
  }>
}

export const surveysApi = {
  list(params: PageParams & { status?: 'active' | 'all' }) {
    return http.get<PageResult<Survey>>('/surveys', { params })
  },

  getById(id: number) {
    return http.get<Survey>(`/surveys/${id}`)
  },

  create(data: SurveyCreate) {
    return http.post<Survey>('/surveys', data)
  },

  update(id: number, data: SurveyUpdate) {
    return http.put<Survey>(`/surveys/${id}`, data)
  },

  submit(id: number, data: SurveySubmit) {
    return http.post<{ message: string }>(`/surveys/${id}/submit`, data)
  },

  getStatistics(id: number) {
    return http.get<SurveyStatistics>(`/surveys/${id}/statistics`)
  },
}
